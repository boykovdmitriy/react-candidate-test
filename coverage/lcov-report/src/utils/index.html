<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for src/utils</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> src/utils
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">39.86% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>57/143</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">27.87% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>17/61</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">37.5% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>18/48</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">41.98% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>55/131</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="api.js"><a href="api.js.html">api.js</a></td>
	<td data-value="4.76" class="pic low"><div class="chart"><div class="cover-fill" style="width: 4%;"></div><div class="cover-empty" style="width:96%;"></div></div></td>
	<td data-value="4.76" class="pct low">4.76%</td>
	<td data-value="21" class="abs low">1/21</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="6" class="abs low">0/6</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="11" class="abs low">0/11</td>
	<td data-value="5.26" class="pct low">5.26%</td>
	<td data-value="19" class="abs low">1/19</td>
	</tr>

<tr>
	<td class="file high" data-value="defAction.js"><a href="defAction.js.html">defAction.js</a></td>
	<td data-value="87.5" class="pic high"><div class="chart"><div class="cover-fill" style="width: 87%;"></div><div class="cover-empty" style="width:13%;"></div></div></td>
	<td data-value="87.5" class="pct high">87.5%</td>
	<td data-value="8" class="abs high">7/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="80" class="pct high">80%</td>
	<td data-value="5" class="abs high">4/5</td>
	<td data-value="85.71" class="pct high">85.71%</td>
	<td data-value="7" class="abs high">6/7</td>
	</tr>

<tr>
	<td class="file low" data-value="defReducer.js"><a href="defReducer.js.html">defReducer.js</a></td>
	<td data-value="25" class="pic low"><div class="chart"><div class="cover-fill" style="width: 25%;"></div><div class="cover-empty" style="width:75%;"></div></div></td>
	<td data-value="25" class="pct low">25%</td>
	<td data-value="12" class="abs low">3/12</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="33.33" class="pct low">33.33%</td>
	<td data-value="3" class="abs low">1/3</td>
	<td data-value="27.27" class="pct low">27.27%</td>
	<td data-value="11" class="abs low">3/11</td>
	</tr>

<tr>
	<td class="file low" data-value="defSaga.js"><a href="defSaga.js.html">defSaga.js</a></td>
	<td data-value="4.17" class="pic low"><div class="chart"><div class="cover-fill" style="width: 4%;"></div><div class="cover-empty" style="width:96%;"></div></div></td>
	<td data-value="4.17" class="pct low">4.17%</td>
	<td data-value="24" class="abs low">1/24</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="5" class="abs low">0/5</td>
	<td data-value="4.17" class="pct low">4.17%</td>
	<td data-value="24" class="abs low">1/24</td>
	</tr>

<tr>
	<td class="file medium" data-value="pageObjectBase.js"><a href="pageObjectBase.js.html">pageObjectBase.js</a></td>
	<td data-value="67.16" class="pic medium"><div class="chart"><div class="cover-fill" style="width: 67%;"></div><div class="cover-empty" style="width:33%;"></div></div></td>
	<td data-value="67.16" class="pct medium">67.16%</td>
	<td data-value="67" class="abs medium">45/67</td>
	<td data-value="54.84" class="pct medium">54.84%</td>
	<td data-value="31" class="abs medium">17/31</td>
	<td data-value="68.42" class="pct medium">68.42%</td>
	<td data-value="19" class="abs medium">13/19</td>
	<td data-value="74.58" class="pct medium">74.58%</td>
	<td data-value="59" class="abs medium">44/59</td>
	</tr>

<tr>
	<td class="file low" data-value="params.js"><a href="params.js.html">params.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="100" class="pct high">100%</td>
	<td data-value="0" class="abs high">0/0</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="1" class="abs low">0/1</td>
	</tr>

<tr>
	<td class="file low" data-value="timeHelpers.js"><a href="timeHelpers.js.html">timeHelpers.js</a></td>
	<td data-value="0" class="pic low"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="8" class="abs low">0/8</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="4" class="abs low">0/4</td>
	<td data-value="0" class="pct low">0%</td>
	<td data-value="10" class="abs low">0/10</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Tue Apr 23 2019 23:55:33 GMT+0300 (Moscow Standard Time)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
