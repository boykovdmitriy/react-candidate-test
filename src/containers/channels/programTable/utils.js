import {
  CHANNEL_ITEM_WIDTH,
  MINUTES_TO_TABLE_PX, TABLE_ROW_HEIGHT,
  TIME_STAMP_BODY_WIDTH,
  TIME_STAMP_HEADER_WIDTH
} from '../../../constants';

const TIME_STAMP_OFFSET = CHANNEL_ITEM_WIDTH + (TIME_STAMP_HEADER_WIDTH - TIME_STAMP_BODY_WIDTH) / 2;

export function calculateTimeStampBodyMetrics(currentTime, channelsCount) {
  return {
    position: currentTime * MINUTES_TO_TABLE_PX + TIME_STAMP_OFFSET,
    bodyHeight: TABLE_ROW_HEIGHT * channelsCount
  }
}

export function calculateNowPosition(currentTime, tableWidth) {
  return currentTime * MINUTES_TO_TABLE_PX + CHANNEL_ITEM_WIDTH - tableWidth/2;
}

export function calculateTimeStampHeaderMetrics(currentTime) {
  const minutes = currentTime % 60;
  const centeringOffset = (TIME_STAMP_HEADER_WIDTH - TIME_STAMP_BODY_WIDTH) / 2;
  return {
    headerPosition: MINUTES_TO_TABLE_PX * minutes,
    bodyPiecePosition: MINUTES_TO_TABLE_PX * minutes + centeringOffset,
  }
}

export function calculateProgramWidth(duration) {
  return duration * MINUTES_TO_TABLE_PX;
}
