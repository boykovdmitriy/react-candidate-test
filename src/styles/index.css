@import "~font-awesome/css/font-awesome.css";
@import "./variables.css";

body {
    margin: 0;
    padding: 0;
    background-color: black;
    overflow: hidden;
}