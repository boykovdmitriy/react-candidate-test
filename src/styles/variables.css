:root {
    --header-height: 3rem;
    --header-top-margin: 1rem;
    --footer-height: 3rem;
    --icon-size: 1.5rem;

    --now-button-width: 3rem;
    --now-button-height: 2rem;
    --now-font-size: 1rem;

    --card-size: 64px;

    --time-slot-width: 240px;

    /*colors*/
    --color-nero: #202020;
    --color-tulip-tree: #E0A131;
    --color-mortar: #545454;
    --color-acadia: #999999;
}