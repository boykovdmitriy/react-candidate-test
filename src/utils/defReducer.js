import {get} from 'lodash';

const initialState = {
  error: '',
  isError: false,
  isLoading: false,
  isLoaded: false,
  data: {},
};

const doNothing = (data) => data;

export function defReducer(actions, config) {
  return (state = initialState, action) => {
    switch (action.type) {
      case actions.REQUEST:
        return {
          ...state,
          data: {},
          isError: false,
          isLoading: true,
          isLoaded: false,
        };
      case actions.SUCCESS:
        const compensate = get(config, 'compensate', doNothing);
        const data = compensate(action.payload.data);
        return {
          ...state,
          data: data,
          isError: false,
          isLoading: false,
          isLoaded: true,
        };
      case actions.FAILURE:
        return {
          ...state,
          isError: true,
          error: action.payload,
          isLoading: false,
          isLoaded: false,
        };
      default:
        return state;
    }
  }
}
