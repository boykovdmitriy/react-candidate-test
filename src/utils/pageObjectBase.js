import React from 'react';
import _ from 'lodash';
import { shallow, mount } from 'enzyme';

const RENDER_STRATEGIES = Object.freeze({
  shallow: (Component, props) => shallow(<Component {...props} />),
  mount: (Component, props) => mount(<Component {...props} />),
});

export class BasePageObject {
  constructor({
                props = {},
                renderStrategy,
              } = {}) {
    this.instanceMocks = [];
    this.propMocks = [];
    this.mocks = {};
    this.customProps = props;
    this.renderStrategy = renderStrategy;
    this.elements = {};

    this.$$register();
    this.data = Object.freeze(this.$$initData());
    this.__buildPropCassette();
    this.__bindPropMocks();
    this.__render();
    this.__bindInstanceMocks();
  }

  $$register() {
    throw new Error('implement $$register hook in your pageObject class');
  }

  $$initData() {
    return {};
  }

  instance() {
    if (this.instanceSelector) {
      return this.tree.find(this.instanceSelector).instance();
    }
    return this.tree.instance();
  }

  setCustomInstanceProps(customProps) {
    this.instance().props = { ...this.instance().props, ...customProps };
  }

  defElement(name, selector) {
    if (this.elements[name]) {
      throw new Error('this element is already defined, name should be uniq');
    }
    this.elements[name] = () => this.tree.find(selector);
  }

  setDefaultRenderStrategy(renderStrategy) {
    if (this.renderStrategy) { return; }
    this.renderStrategy = renderStrategy;
  }

  setComponent(component, instanceSelector) {
    if (this.component) { throw new Error('component already registered'); }
    this.component = component;
    this.instanceSelector = instanceSelector;
  }

  setDefaultPropCassette(cassette) {
    if (this.defaultPropCassette) { throw new Error('defaultPropCassette already defined'); }
    this.defaultPropCassette = cassette;
  }

  mockProp(name, callback = _.noop) {
    if (this.mocks[name]) { throw new Error('mock is already defined'); }
    this.propMocks.push({ name, callback });
  }

  __buildPropCassette() {
    const cassette = this.defaultPropCassette;
    if (!cassette) { throw new Error(`no default cassette`); }
    this.props = Object.freeze(cassette(this.data));
  }

  __bindPropMocks() {
    const props = this.propMocks.reduce((acc, { name: key, callback }) => {
      this.mocks[key] = jest.fn(callback);
      acc[key] = this.mocks[key];
      return acc;
    }, {});

    this.props = { ...this.props, ...props };
  }

  __bindInstanceMocks() {
    if (this.renderStrategy === 'shallow') {
      if (!_.isEmpty(this.instanceMocks)) {
        throw new Error('You can not mock props when you are using SHALLOW render strategy');
      }
      return;
    }
    if (!this.instanceMocks.length) { return; }
    const instanceProps = this.instanceMocks.reduce((acc, key) => {
      this.mocks[key] = jest.fn(this.instance().props[key]);
      acc[key] = this.mocks[key];
      return acc;
    }, {});
    this.setCustomInstanceProps(instanceProps);
  }

  __render() {
    if (!this.component) { throw new Error('define component in  already $$register hook'); }
    if (!this.renderStrategy) {
      throw new Error('define render strategy either via setDefaultRenderStrategy in $register hook of via contructor passing in renderStrategy key.');
    }
    const strategy = RENDER_STRATEGIES[this.renderStrategy];
    if (!strategy) {
      throw new Error(`passed strategy is not supported ${this.renderStrategy}. Pass one of: ${Object.keys(RENDER_STRATEGIES)}`);
    }
    const customProps = _.isFunction(this.customProps)
      ? this.customProps(this.data)
      : this.customProps;
    this.tree = strategy(
      this.component,
      { ...this.props, ...customProps },
    );
  }
}
