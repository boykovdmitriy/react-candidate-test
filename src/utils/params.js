import qs from 'qs';

export function toParams(object) {
  return qs.stringify(object, { addQueryPrefix: true });
}
