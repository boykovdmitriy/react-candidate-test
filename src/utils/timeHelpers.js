import moment from 'moment';

export const SERVER_TIME_FORMAT = 'YYYY-MM-DD HH:mm:ss';

export function calculateDuration(startTime, endTime) {
  if (!startTime || !endTime) {
    return 0;
  }
  const start = moment.isMoment(startTime) ? startTime : moment(startTime, SERVER_TIME_FORMAT);
  const end = moment.isMoment(endTime) ? endTime : moment(endTime, SERVER_TIME_FORMAT);
  return end.diff(start, 'minutes');
}

export function getCurrentTimeInMinutes() {
  const date = new Date();
  return date.getHours() * 60 + date.getMinutes();
}

export function getTimeInMinutes(time) {
  const date = moment(time, SERVER_TIME_FORMAT);
  return date.hours()*60 + date.minutes();
}

export function getCurrentWeekday() {
  return moment(new Date()).format('ddd');
}
