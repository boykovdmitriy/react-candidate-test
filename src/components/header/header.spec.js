import cx from 'classnames';
import {Header} from './index';
import styles from './header.css';
import {BasePageObject} from '../../utils/pageObjectBase';

class PageObject extends BasePageObject {
  $$register() {
    this.setComponent(Header, 'Header');
    this.setDefaultRenderStrategy('shallow');
    this.setDefaultPropCassette(({text, className}) => ({
      children: text,
      className
    }));
  }

  $$initData() {
    return {
      text: 'some useful information',
      className: 'customClassName'
    }
  }

  didRenderChildren() {
    const {text} = this.data;
    expect(this.tree.text()).toEqual(text);
  }

  didRenderWithClassName() {
    const {className} = this.data;
    expect(this.tree.props().className).toEqual(cx(styles.container, className));
  }
}

describe('Header', () => {
  it('renders correctly', () => {
    const pageObject = new PageObject();
    pageObject.didRenderChildren();
    pageObject.didRenderWithClassName();
    expect(pageObject.tree).toMatchSnapshot();
  });
});
