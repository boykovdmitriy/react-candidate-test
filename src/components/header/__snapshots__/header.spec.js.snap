// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders correctly 1`] = `
<header
  className="container customClassName"
>
  some useful information
</header>
`;
